# nibscbioinformatics/viralevo: Output

This document describes the output produced by the pipeline. Results are reported in the generated HTML file analysis_report.html and QC information is presented in a MultiQC report.

## Pipeline overview

The pipeline is built using [Nextflow](https://www.nextflow.io/)
and processes data using the following steps:

* [FastQC] - read quality control
* [MultiQC] - aggregate QC report
* [CutAdapt] - trimming for quality and adapter sequence
* [BWA] - alignment to reference
* [SamTools] - processing alignment
* [LoFreq] - variant calling
* [iVar] - variant calling
* [MUSCLE] - multiple sequence alignment of consensus
* [jModelTest2] - phylogenetic tree with samples
* [SPADES] - de novo assembly

Full details of the software used and the methods are provided in the analysis report generated by the pipeline, and in analysis_report.Rmd.

## Output folder structure

On running the pipeline, users must specify a root output directory, with --outdir /my/output/directory within which the pipeline creates subdirectories for results. The following subdirectories are created within the user-specified output folder:

* [alignments] - contains the depth table, and a subfolder for each sample containing aligned BAM files, consensus sequence and a SPADES output folder with de novo assembly files
* [calling] - contains variant caller output from LoFreq and iVar as well as an aggregate variant table
* [fastqc] - contains output from FastQC
* [MultiQC] - contains the MultiQC report
* [phylogenetic] - contains MUSCLE and jModelTest output files
* [pipeline_info] - automatically generated pipeline information
* [reports] - contains the analysis report
* [stats] - contains tables summarising alignment and trimming logs

* [work] - unless otherwise specified with -w /my/work/folder this will be placed in the working directory, and contain Nextflow working files from the run
